packages/twoside/init.lua

```diff
@@ -65,28 +66,74 @@ function package:_init (options)
   self.class:registerPostinit(function (class)
     class:mirrorMaster(options.oddPageMaster, options.evenPageMaster)
   end)
+  self.class:registerHook("newpage", spreadHook)
   self.class:registerHook("newpage", switchPage)
 end
 
 function package:registerCommands ()
 
   self:registerCommand("open-double-page", function()
-    SILE.typesetter:leaveHmode()
+    SILE.call("open-spread", { double = false, odd = true, blank = false })
+  end)
+
+  self:registerCommand("open-spread-eject", function()
     SILE.call("supereject")
-    if self.class:oddPage() then
-      SILE.typesetter:typeset("")
-      SILE.typesetter:leaveHmode()
-      SILE.call("supereject")
+  end)
+
+  -- This is upstreamed from CaSILE. Similar to the original open-double-page,
+  -- but can disable headers and folios on blank pages and allows opening the
+  -- even side (with or without a leading blank).
+  self:registerCommand("open-spread", function (options)
+    local odd = SU.boolean(options.odd, true)
+    local double = SU.boolean(options.double, true)
+    local blank = SU.boolean(options.blank, true)
+    local optionsMet = function ()
+      return (not double or spread_counter > 1) and
+             (odd and self.class:oddPage()) or
+             (not odd and not self.class:oddPage())
     end
+    spread_counter = 0
     SILE.typesetter:leaveHmode()
+    -- Output a box, then remove it otherwise we can't prove what page new content will land on
+    SILE.call("hbox"); SILE.typesetter:leaveHmode(); table.remove(SILE.typesetter.state.nodes)
+    if spread_counter == 1 and optionsMet() then
+      SILE.typesetter.state.outputQueue = {}
+      return
+    end
+    repeat
+      if spread_counter > 0 then
+        SILE.call("hbox")
+        SILE.typesetter:leaveHmode()
+        if blank then
+          SILE.scratch.headers.skipthispage = true
+          SILE.scratch.counters.folio.off = 2
```

Comment:
Again breaking separation of concerns. We have `\nofolios`, `\nofoliothispage` etc. but this package hits directly into scratch variable. Maybe my soon-to come re-implementation of folios won't have this scratch variable, though be compatible at command level? How much internal knowledge of scratch things has to be know by class & packages designers?

Reply:
Good call on using the public facing folios command. I switched to that since it exists. When we have a book class with proper public API for handling headers we can switch to that too, but we don't right yet and it's out of scope for this PR.